### DISCUSSION TITLE
Head bubbling

### DISCUSSION THREAD

**Discussion by matthewp** [withastro/roadmap#481] [Proposal]
### Body

# Summary

Allow components to define a `<head>` section which will bubble up to the head of the page.

# Terms

These terms are used through the doc, so I'm defining them here:

- **Head injection** - The process where Astro adds `<link>`, `<style>` and `<script>` tags to the end of the `<head>` section for your page's styles and hoisted scripts.
- **Implicit head injection** - The process where Astro tries to insert into the head even if you do not define an explicit `<head>` tag in any components.
- **Head propagation** - The process where Astro "bubbles" up head content from child components. Today this only happens through Content Collections, to add links and styles only when you render the `Content` component.

# Background & Motivation

This RFC seeks to solve 2 core problems:

1. Implicit head injection is difficult to implement give that components are not guaranteed to run in top-down order. This problem has always existed in Astro but has become harder with JSX components and head propagation in content collections. This is the source of bugs and a time sink.
1. Data loading patterns in Astro today need to be done at the page or layout level, so the data can be used for both head and body content. Head bubbling would allow data loading to be done within components, which makes reuse easier.

The background for why to do this, and why now:

- Head propagation became possible at the low-levels of Astro with Content Collections.
- Some other frameworks / libraries, such as Svelte and react-helmet have this ability and it's quite popular there.
- Users have asked for head bubbling since the early days of Astro.

# Goals

- Deprecate implicit head injection, making it a requirement that your page has explicit `<html>` and `<head>` tag in some component (could be page or layout).
- Allow components to define `<head>` sections and have Astro wait on these components to run before streaming the head section.
- Preserve streaming as a default of Astro; head propagation blocks but only if you use it, and it only blocks on components using it, not on the entire page.
- Provide a mechanism for slotting head content into a certain *position* in the head. For example some head content needs to happen at the first.

# Example

```astro
---
const food = await fetch('http://external.api/food.json?id=' + Astro.props.id);
---

<head>
  <title>{ food.name }</title>
  <meta property="og:title" content={ food.name } />
  <meta property="og:description" content={ food.longDescription } />
  <meta property="og:image" content={ food.image } />
</head>

<h1>{ food.name }</h1>
```

# Details / Notes

- The `<head>` tag **must** be top-level in the component to be hoisted. If it is nested, such as `<html><head>` then it is assumed that the intent is to print out the head tag itself.

**Comment by wassfila** [comment]
Some in support posts were asking for this actually ( I think used to it from other frameworks). The alternative of using slots in layout is not easy for all skill levels, and is from page only and not from components, for a good reason => My opinion is that it's an anti-pattern and I do not think the users will realize the price to pay "Astro wait on these components to run before streaming the head section".
Risk to play against Astro benchmarks Mabye be enabled by an explicit flag ? or force disabled with a performance flag to be sure?

**Reply by matthewp** [reply]
As far as flags to turn this on/off, there will need to be a flag at first as this will be a breaking change. 

Also should note that if you do not use head bubbling there is no cost in streaming. The head propagation mechanism knows if there is a child in the tree that does bubbling and will only wait if there's such a component.

**Reply by bholmesdev** [reply]
> My opinion is that it's an anti-pattern and I do not think the users will realize the price to pay

I'd push back on this a bit. Performant per-component streaming is a _very_ nuanced case that newer users shouldn't worry about. It sounds like this would block in the same way an explicit `slot` would block, which feels acceptable to me.

**Comment by tony-sull** [comment]
Very exciting to see this one come to Astro, and with a plain old `<head>` tag instead of `<astro:head>` no less!

Would there be any kind of check for duplicate tags?  ex: two components add a `<title>` or `<meta property="og:description">` tag

**Reply by matthewp** [reply]
No, for a few reasons:

- In `.astro` components HTML is compiled to a template string. We do not have an easy way to deduplicate its contents.
- I don't think we could deduplicate deterministically, because components do not run in top-down order (JSX runs eagerly, as do slots), so we would wind up with the same sort of bugs that we have with implicit head injection. One of the goals of this proposal is to get rid of this.
- We would want to eventually create an API for renderers to do head bubbling as well, and they always render to a string so we can't deduplicate without HTML parsing.

**Reply by smnbbrv** [reply]
@matthewp thanks! My crying was heard 😂

Is there a possibility to create hooks for e.g. 

- telemetry (to measure the lag caused by head awaiting), e.g to connect with opentelemetry / metrics for prometheus?
- head postprocessor (as a replacement for requested above for custom content deduplication)

Ideally this is not limited to the header only

**Reply by matthewp** [reply]
Both of those things would be out of scope for this proposal. Any sort of hook for postprocessing would just be handing you a string of HTML anyways. So you'd have to parse it and manipulate it. Not something we should encourage, and something that can already be done using the normal integration hooks.

**Reply by bholmesdev** [reply]
Agreed, really like we aren't using a special `<astro:head>` tag! This feels inline with our magic `style` and `script` tags

**Comment by bluwy** [comment]
> * Deprecate implicit head injection, making it a requirement that your page has explicit `<html>` and `<head>` tag in some component (could be page or layout).

If we implement the special rule for where `<head>` can be define in subtree components, wouldn't that solve the implicit head injection issue? That way we could preserve implicit head injection, by after collecting all the heads and there's no existing `<head>`, we could inject one ourselves.

**Reply by matthewp** [reply]
Just to clarify, implicit head injection means that if you're entire page is like this:

```astro
<div>Hello world</div>
```

We insert head contents before that `<div>`, resulting in something like:

```astro
<link rel="stylesheet" href="/bundle/foo.css">
<div>Hello world</div>
```

If you see references to `maybeRenderHead` in the JS output, that's what this is for.

But this is problematic and the source of bugs, especially with the introduction of content collections which enables head propagation. The reason is that implicit head injection works best when everything runs in top-down depth-first order. But that's not the case in reality, which means some times we can inject the head contents into the wrong place.

So a big part of this proposal is removing that, and forcing you to have this somewhere in your page:

```astro
<html>
  <head>
    ...
  </head>
```

As long as you have that, we will always insert the head contents into the right place.

**Reply by bluwy** [reply]
I think I understand now, if we don't have an explicit head tag, things running in a non-top-down order might inject the head at the wrong spot, and would be especially complicated for streaming.

I think I can get by a warning if no `head` tag is defined, but maybe still have the fallback best-effort strategy for now so we don't need to error out when no `head` tag is defined.

**Comment by Princesseuh** [comment]
> to the end of the <head> section for your page's styles and hoisted scripts.

Any thoughts on being able to control the order in some ways? We've had requests in the past to be able to add certain tags at a certain position in the resulting `head` (presumably to do some over-the-top optimization regarding certain tags being first). Something like a `<head inject='first'>` perhaps.

I also assume that the order in which things are added is not deterministic, and as such adding things like CSS can result in different styles depending on which components finished first? Is this something we would warn about doing?

**Reply by jamesarosen** [reply]
Yes! And if your server doesn't give you control over HTTP headers, using `<meta http-equiv>` is often a good alternative. Some of those "headers" also have ordering requirements. For example, `Content-Security-Policy` must come before any resource links: favicon, JS, CSS.

**Reply by matthewp** [reply]
Yes, I like this idea a lot. I've added as a **goal** without giving a specific API, that will have to figure out.

There's some overlap here with the concept of slots, so maybe that's a mechanism to resume? I could see something like this:

__layout.astro__

```astro
<html>
  <head>
    <slot name="top">
    
    <title>...</title>
  </head>
</html>
```

__child.astro__

```astro
<head slot="top">
  <meta http-equiv ...>
</head>
```

Just an idea, there might be some reasons not to reuse the slot name.

**Comment by jamesarosen** [comment]
This seems related because of the change to head-streaming, but it's not exactly the same:

I can modify the `Response` in a page:
```
// src/pages/index.astro
Astro.response.headers.set('from-page', 'foo') // 😀
```

But not in a layout:
```
// src/layouts/main.astro
Astro.response.headers.set('from-layout', 'bar') // 😞
```

Or from a component:
```
// src/components/sidebar.astro
Astro.response.headers.set('from-component', 'baz') // 😞
```

I mention it because many of the reasons for injecting `<head>` content overlap with reasons for setting Response headers. Obviously the APIs aren't the same, but some of the underlying architectural changes might be.

**Reply by smnbbrv** [reply]
Not only headers, the status code could be also something to change. 

Example:

Product list page with pagination, if there are no products on page 2, semantically correct would be to return 404. 

This information most likely is available on product list component only. Current astro would require resolve this data on a page level which is not really comfortable and not always possible (e.g. if page content comes from external source / CMS)

Not that common, but happens

**Reply by matthewp** [reply]
I was thinking that this is probably possible when I wrote the proposal. If we wait on `<head>` then by nature it should also be possible to allow setting headers and setting the status code. Ultimately I left it out of the proposal because I wasn't sure what the API for that should be. With `<head>` we can statically mark the component as needing to be awaited on; but for modifying headers, since that happens in JS we don't have a good mechanism that I can think of. So maybe this is something we can figure out in implementation.

**Reply by natemoo-re** [reply]
This is a particularly compelling use case to me, I'd definitely want to explore this more. Streaming and non-200 responses are typically thought of as incompatible, but if we could solve this it would open up a ton of use cases.

**Reply by matthewp** [reply]
Honestly, it "just works" as long as you have a head. You could even have an empty head:

```astro
---
Astro.response.headers.set('X-Foo', 'bar');
---

<head></head>

<div>Hello world</div>
```

This will block on head buffering and the headers are set, even though you don't have any real head content. I almost feel like this a good-enough solution to recommend it.

**Reply by jamesarosen** [reply]
So cool!

I might do something like this in my app, just to improve the cohesion:
```astro
---
// src/components/ResponseHeaders.astro
Object.entries(Astro.props).forEach([name, value] => {
  Astro.response.headers.set(name, value);
})
---
<head></head>
```

**Reply by matthewp** [reply]
@jamesarosen Yes, that will work with the current preview release. Returning a `Response` from the frontmatter will *not* at the moment. But I think it's something we can fix relatively easily if this proposal is accepted.

**Comment by matthewp** [comment]
Implementation was done here: https://github.com/withastro/astro/pull/6286. Will be revisiting this idea soon, hopefully.

**Comment by nadar** [comment]
I would love to see this feature moving forward. I see it from an "integrations" perspective, when the ecosystem should grow even more, this is an essential function for CMS and headless systems which have, for example, meta information but are not able to automatically inject those data into the user's layout.
